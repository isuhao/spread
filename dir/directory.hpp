#ifndef __SPREAD_DIRECTORY_HPP_
#define __SPREAD_DIRECTORY_HPP_

#include <map>
#include <string>
#include <json/json.h>
#include "hash/hash.hpp"
#include <mangle/stream/stream.hpp>

namespace Spread
{
  struct Directory
  {
    typedef std::map<std::string,Hash> DirMap;
    DirMap dir;

    // Read and write JSON data
    void readJson(const std::string &file);
    void parseJson(const Json::Value &val);
    void writeJson(const std::string &file) const;
    Json::Value makeJson() const;

    /* Read and write binary format. Returns the hash of the data read
       or written.
     */
    Hash read(const std::string &file);
    Hash read(Mangle::Stream::StreamPtr strm);
    Hash write(const std::string &file) const;
    Hash write(Mangle::Stream::StreamPtr strm) const;

    // Find a member entry. Returns a null hash if not found.
    const Hash &find(const std::string &name) const;
  };
}

#endif
