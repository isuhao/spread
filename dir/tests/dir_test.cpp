#include "directory.hpp"
#include <iostream>

using namespace std;
using namespace Spread;

int main()
{
  Directory dir;

  dir.readJson("test.json");
  cout << dir.find("name1") << endl;
  cout << dir.find("name2") << endl;
  cout << dir.find("dir/name2") << endl;
  dir.write("test.bin");

  Directory dir2;
  dir2.read("test.bin");
  cout << dir2.makeJson();

  return 0;
}
