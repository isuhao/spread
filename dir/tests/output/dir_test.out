|~X[UfCU[9a4aHmAQ7)1-,4l,x<YL=^CfdZ*]>;K5^d
00
L).Z-5TI7:AJ{l+Mk6qi}.oNG{Yh^FH007w3noK)M*

{
	"dir/name2" : "L).Z-5TI7:AJ{l+Mk6qi}.oNG{Yh^FH007w3noK)M*",
	"name1" : "|~X[UfCU[9a4aHmAQ7)1-,4l,x<YL=^CfdZ*]>;K5^d"
}
